### PR TITLE
Use Difference operator for discrete system

### PR DESCRIPTION
**Pull request by sharanry**
https://github.com/SciML/ModelingToolkit.jl/issues/1088#issuecomment-876649631

TODO:

- [x] Add utilities which enable handling Difference operators.
- [x] Add difference function generators
- [x] Add tests

### CODE REVIEW

**Comment by sharanry** [comment]
@ChrisRackauckas @YingboMa The simple SIR test now works fine Difference operator. Is there any other more complex scenarios I can look at and support or is this sufficient for this initial PR?

**Comment by sharanry** [comment]
Also, I think currently FunctionMap defaults `dt=1` [here](https://github.com/SciML/OrdinaryDiffEq.jl/blob/ba2a3a841614182a75c37e9e48b2c9ee82c60a0a/src/solve.jl#L26). I guess we will have to change that?

**Comment by sharanry** [comment]
One other thing we are assuming right now is that we have a common difference operator on all states with equal timesteps. Do we intend to support multi-timestep processes?

**Comment by ChrisRackauckas** [comment]
>Also, I think currently FunctionMap defaults to dt=1 here. I guess we will have to change that?

Yes.

File: src/systems/discrete_system/discrete_system.jl [hunk 1]
```diff
@@ -115,8 +115,82 @@ function DiffEqBase.DiscreteProblem(sys::DiscreteSystem,u0map,tspan,
     u = dvs
     p = varmap_to_vars(parammap,ps)
 
-    f_gen = build_function(rhss, dvs, ps, t; expression=Val{eval_expression}, expression_module=eval_module)
-    f_oop,f_iip = (@RuntimeGeneratedFunction(eval_module, ex) for ex in f_gen)
+    f_gen = generate_function(sys; expression=Val{eval_expression}, expression_module=eval_module)
+    f_oop, _ = (@RuntimeGeneratedFunction(eval_module, ex) for ex in f_gen)
     f(u,p,t) = f_oop(u,p,t)
```
Comment (YingboMa) [inline src/systems/discrete_system/discrete_system.jl:120]: Shouldn't there also be an in-place version as well?

File: src/systems/discrete_system/discrete_system.jl [hunk 2]
```diff
@@ -115,8 +115,82 @@ function DiffEqBase.DiscreteProblem(sys::DiscreteSystem,u0map,tspan,
...
     DiscreteProblem(f,u0,tspan,p;kwargs...)
 end
+
+isdifference(expr) = istree(expr) && operation(expr) isa Difference
+isdifferenceeq(eq) = isdifference(eq.lhs)
+
+difference_vars(x::Sym) = Set([x])
```
Comment (YingboMa) [inline src/systems/discrete_system/discrete_system.jl:127]: Do you really need to copy-paste the `vars` code? What did you change?

File: src/systems/discrete_system/discrete_system.jl [hunk 3]
```diff
@@ -115,8 +115,82 @@ function DiffEqBase.DiscreteProblem(sys::DiscreteSystem,u0map,tspan,
...
+
+    symtype(operation(O)) <: FnType && push!(difference_vars, O)
+    for arg in arguments(O)
+        difference_vars!(difference_vars, arg)
+    end
+
+    return difference_vars
+end
+
+function collect_difference_variables(sys::DiscreteSystem)
```
Comment (YingboMa) [inline src/systems/discrete_system/discrete_system.jl:153]: Let's try not duplicate too much code. This code could be easily reused by making it a higher order function and pass in `isdifference`.

**Review by YingboMa — CHANGES_REQUESTED** [review]
The code copy-pasting seems easily avoidable.

File: src/systems/discrete_system/discrete_system.jl (hunk 1 above)
Comment (sharanry) [reply src/systems/discrete_system/discrete_system.jl:120]: Yes. It does get generated. But not sure if we have any use for it in discrete systems as even solvers like `FunctionMap` uses the not-in-place function (not sure what it is called 😅 ).

**Comment by sharanry** [comment]
Yup. I will try and refactor the code.

**Comment by sharanry** [comment]
@YingboMa @ChrisRackauckas Does the ODESystem assume some kind of ordering of the equations in the differential system? It looks like the `generate_function` for ODESystem does not take into consideration what the LHS of each equation is. I also noticed [this TODO comment](https://github.com/SciML/ModelingToolkit.jl/blob/master/src/systems/diffeqs/abstractodesystem.jl#L112).

**Review by YingboMa — APPROVED** [review]
LGTM

**Comment by YingboMa** [comment]
Oh, remember to also generalize the `vars!` function.

**Comment by sharanry** [comment]
> Discrete system doesn't require it. Will add it in the difference operator in ODE system PR on which I have started a discussion [here](https://github.com/SciML/ModelingToolkit.jl/issues/1122)?

**Comment by YingboMa** [comment]
Yes, handling `vars!` will be required for that issue, but let's do that in a future PR.

**Pull request merged**